src/main/java/backend/resource/MultiModel.java

```diff
@@ -201,21 +202,33 @@ private void preprocessUpdatedIssues(List<Model> newModels) {
         for (Model model : newModels) {
             assert models.containsKey(model.getRepoId());
             Model existingModel = models.get(model.getRepoId());
-            if (!existingModel.getIssues().equals(model.getIssues())) {
-                // Find issues that have changed and update preferences with them
-                for (int i = 1; i <= model.getIssues().size(); i++) {
-                    // TODO O(n^2), optimise by preprocessing into a map or sorting
-                    if (!existingModel.getIssueById(i).equals(model.getIssueById(i))) {
-                        assert model.getIssueById(i).isPresent();
-                        // It's no longer currently read, but it retains its updated time.
-                        // No changes to preferences.
-                        model.getIssueById(i).get().setIsCurrentlyRead(false);
-                    }
-                }
+            if (existingModel.getIssues().equals(model.getIssues())) continue;
+            for (int i = 1; i <= model.getIssues().size(); i++) {
+                if (existingModel.getIssueById(i).equals(model.getIssueById(i)) || 
+                    !wasNotJustMarkedAsRead(model.getIssueById(i).get())) continue;
+                // TODO O(n^2), optimise by preprocessing into a map or sorting
+                assert model.getIssueById(i).isPresent();
+                /** If the issue was marked read after the last update, but update came in later
+                 * due to network latency, ensure the issue retains its read state
+                **/
+                model.getIssueById(i).get().setIsCurrentlyRead(false);
             }
         }
     }
 
+    private boolean wasNotJustMarkedAsRead(TurboIssue issue) {
+        if (!issue.isCurrentlyRead() || !issue.getMarkedReadAt().isPresent()) return false;
```

Comment:
Maybe it's just me, but I find it quite difficult to reason about this function's correctness still. We should break it up into smaller functions. That would make it easier to test as well.

Something like this? The logic should be equivalent.

```java
/**
 * Determines if an issue was recently marked as read. This happens if
 * it is marked as read a short time before an update occurs.
 */
private boolean wasJustMarkedAsRead() {
    if (!isCurrentlyRead()) return false;
    return wasReadWithinInterval(3L, ChronoUnit.SECONDS);
}

/**
 * Determines if an issue was read in a short interval before its last update.
 */
private boolean wasReadWithinInterval(long seconds, ChronoUnit unit) {
    if (!getMarkedReadAt().isPresent()) return false;
    return getUpdatedAt().isBefore(getMarkedReadAt().get())
        && getUpdatedAt().until(getMarkedReadAt().get(), unit) < seconds;
}
```

I put it inside TurboIssue. Seems to fit better because it's entirely a function of the issue's state.